Add Igalia's blog post to review web browser implementations in 2020 (#502)
ipfs/blog: content/post/115-ipfs-and-igalia-collaborate-on-dweb-in-browsers.md

---
date: 2021-01-15

> url: /115-ipfs-and-igalia-collaborate-on-dweb-in-browsers/

Reviewer: I believe  the convention for non-weekly articles is to include date in the url slug, so its easy to tell how old content is:

```suggestion
url: /2021-01-13-ipfs-and-igalia-collaborate-on-dweb-in-browsers/
```

Author: mmh, there does not seem to be any other posts like that?
FWIW, I followed https://github.com/ipfs/blog#create-a-new-blog-post here.

## Distributed web protocols in custom handlers

<div style="width: 207pt; margin-left: auto; margin-right: auto;">
<svg width="207pt" height="305pt" viewBox="0.00 0.00 207.04 305.00">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 301)">
<polygon fill="transparent" stroke="transparent" points="-4,4 -4,-301 203.04,-301 203.04,4 -4,4"/>
<!-- 0 -->
<g id="node1" class="node">
<title>0</title>
<polygon fill="none" stroke="black" points="103.04,-297 49.04,-297 49.04,-261 103.04,-261 103.04,-297"/>
<text text-anchor="middle" x="76.04" y="-275.3" font-family="Times,serif" font-size="14.00">User</text>
</g>
<!-- 1 -->
<g id="node2" class="node">
<title>1</title>
<polygon fill="none" stroke="black" points="114.54,-210 37.54,-210 37.54,-174 114.54,-174 114.54,-210"/>
<text text-anchor="middle" x="76.04" y="-188.3" font-family="Times,serif" font-size="14.00">Browser</text>
</g>
<!-- 0&#45;&gt;1 -->
<g id="edge1" class="edge">
<title>0&#45;&gt;1</title>
<path fill="none" stroke="black" d="M76.04,-260.8C76.04,-246.05 76.04,-224.92 76.04,-210.18"/>

> <text text-anchor="middle" x="111.54" y="-231.8" font-family="Times,serif" font-size="14.00">ifps://blah</text>

Author: This SVG was generated with `dot`. I'm not sure there is a proper place where to put the source. Anyway, will update the thing

Since there is a strong preference within the Chromium project to rely on Web APIs for new extension features, code owners’ counter-proposal has been to give more power to `registerProtocolHandler` in extension context ([cross-origin handler](https://chromium-review.googlesource.com/c/chromium/src/+/2287304) and [extension-specific schemes](https://chromium-review.googlesource.com/c/chromium/src/+/2560305)). The two patches for these landed recently in Chromium and will be available in version 89!

For now, due to an existing [permission UI bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1079333), the protocol registration must happen in an extension tab, but you can already get an idea of the new possibility [in this video](https://vimeo.com/489748722), (be sure to enable subtitles for a detailed description):

<div style="width: 640px; margin-left: auto; margin-right: auto">
<iframe src="https://player.vimeo.com/video/489748722" width="640" height="361" frameborder="0" allow="autoplay; fullscreen" allowfullscreen></iframe>
</div>

## Secure Contexts

> The issue with redirected URLs not behaving like normal URLs is a bit more tricky and can come from several reasons. For local HTTP gateways, such as the one provided by [IPFS Desktop](https://github.com/ipfs-shipyard/ipfs-desktop#ipfs-desktop), one of the explanations is that browsers used to not treat these local URLs as insecure contexts and thus block various web platform features. This was [changed three years ago in Chrome](https://www.chromestatus.com/feature/6269417340010496) and specifications were updated accordingly. More precisely:

Reviewer: This looks like a typo (double negative?)

```suggestion
The issue with redirected URLs not behaving like normal URLs is a bit more tricky and can come from several reasons. For local HTTP gateways, such as the one provided by [IPFS Desktop](https://github.com/ipfs-shipyard/ipfs-desktop#ipfs-desktop), one of the explanations is that browsers used to not treat these local URLs as secure contexts and thus block various web platform features. This was [changed three years ago in Chrome](https://www.chromestatus.com/feature/6269417340010496) and specifications were updated accordingly. More precisely:
```

Mozilla has been supportive of this change. The case of Loopback IP addresses has been implemented since Firefox 55, but as this happens with limited development resources and prioritization, the work for localhost names had never been finished. One of the difficulties being that many existing network tests do not assume the above behavior and require some adjustments to keep passing. After several attempts, the [main patches landed without being reverted](https://bugzilla.mozilla.org/show_bug.cgi?id=1220810) and is released in the latest version of Firefox.

The position within the WebKit community is less clear, some members being skeptical about granting web sites access to local hosts, others thinking the behavior implemented in Chrome and Firefox make sense. Similarly to Firefox, one difficulty is that many tests need to be tweaked. Additionally, implementing the address redirection does not seem to be easily doable at the WebKit level. One can find details [on the corresponding bug](https://bugs.webkit.org/show_bug.cgi?id=171934). In order to get things rolling, patches following a preference-based approach have been submitted for review.

> This problem is actually more general than just local resources. The whole notion of “secure context” is not implemented consistently between browsers, or even within each browser. In addition to the one of [potentially trustworthy origin](https://w3c.github.io/webappsec-secure-contexts/#is-origin-trustworthy) mentioned above, the specification has a slightly more general notion of [potentially trustworthy URL](https://w3c.github.io/webappsec-secure-contexts/#is-url-trustworthy) which used to differentiate from [a priori authenticated URL](https://github.com/w3c/webappsec-mixed-content/issues/35). How these notions should interact with custom handlers is also [currently a bit fuzzy](https://github.com/whatwg/html/issues/5787https://github.com/whatwg/html/issues/5787).

Reviewer: removed duplicated link:
```suggestion
This problem is actually more general than just local resources. The whole notion of “secure context” is not implemented consistently between browsers, or even within each browser. In addition to the one of [potentially trustworthy origin](https://w3c.github.io/webappsec-secure-contexts/#is-origin-trustworthy) mentioned above, the specification has a slightly more general notion of [potentially trustworthy URL](https://w3c.github.io/webappsec-secure-contexts/#is-url-trustworthy) which used to differentiate from [a priori authenticated URL](https://github.com/w3c/webappsec-mixed-content/issues/35). How these notions should interact with custom handlers is also [currently a bit fuzzy](https://github.com/whatwg/html/issues/5787).
```

> In Chromium, there were [at least 6 implementations of “secure”](https://bugs.chromium.org/p/chromium/issues/detail?id=1153336). Some of these implementations are not very aligned with the current specification. Worse, some of them include historical and non-standard localhost names like `localhost.localdomain`, `localhost6` and `localhost6.localdomain6`. Effort is being made to progressively and carefully unify these implementations and follow the specifications. In particular, patches landed to remove [localhost](https://chromestatus.com/feature/5698580851458048) [names](https://chromestatus.com/feature/5668106045227008) and make more [`data:` URLs potentially trustworthy](https://chromestatus.com/feature/5634194258526208).

Reviewer: small clarifications
```suggestion
In Chromium, there were [at least 6 implementations of “secure”](https://bugs.chromium.org/p/chromium/issues/detail?id=1153336). Some of these implementations are not very aligned with the current specification. Worse, some of them include historical and non-standard localhost names like `localhost.localdomain`, `localhost6` and `localhost6.localdomain6`. Effort is being made to progressively and carefully unify these implementations and follow the specifications. In particular, patches landed to remove non-standard [localhost](https://chromestatus.com/feature/5698580851458048) [names](https://chromestatus.com/feature/5668106045227008) and make [`data:` URLs potentially trustworthy](https://chromestatus.com/feature/5634194258526208).
```

## Other fixes

As usual, performing this kind of effort leads to a lot of side tasks: community and specification discussion, code clean up and refactoring, documentation improvement, and other bug fixes. Here are Chromium bugs fixed that are relevant for custom handlers and interoperability with Firefox:

* [Make service workers work with custom handlers](https://chromium-review.googlesource.com/c/chromium/src/+/2487107)
* [Do not remove %s token when validating (un)registerProtocolHandler’s URL](https://chromium-review.googlesource.com/c/chromium/src/+/2335434)
* [percent-encode U+0020 SPACE when in URLs computed by custom protocol handlers](https://chromestatus.com/feature/5678518908223488)
* [percent-encode the delete character when parsing URLs](https://chromestatus.com/feature/5651438652882944)

Finally, for both Chromium and Firefox, [the title argument has been removed from registerProtocolHandler()](https://chromestatus.com/feature/5634386663833600) as per [Mozilla's suggestion](https://github.com/whatwg/html/pull/5425).

> If you are interested in an exhaustive list of commits, try expanding the following list:
> 
> <details><summary>Chromium</summary>

Reviewer: Our blog engine does not support collapsed sections. I think we should just show the list, to emphasize how complex and involved this work was:

```suggestion
If you are interested in an exhaustive list of contributions related to this work, here it comes:

###  Chromium
```

* [2617709: Rewrite SecurityOrigin::IsSecure() using GURL and url::Origin](https://chromium-review.googlesource.com/c/chromium/src/+/2617709)
* [2617883: Make SecurityOrigin::IsSecure treat localhost and local files as secure](https://chromium-review.googlesource.com/c/chromium/src/+/2617883)
</details>

<details><summary>Firefox</summary>

* [Improve compatibility of protocol_handlers with registerProtocolHandler](https://phabricator.services.mozilla.com/D71747)
* [Safelist cabal, dat, did, dweb, ethereum, hyper, ipfs, ipns, and ssb schemes for registerProtocolHandler().](https://phabricator.services.mozilla.com/D71580)
* [Remove the title argument from registerProtocolHandler()](https://phabricator.services.mozilla.com/D71714)
* [ Do not set network.dns.ipv4OnlyDomains when running XPCShell](https://phabricator.services.mozilla.com/D94986)
* [Add a test to ensure loopback host names cannot be overridden](https://phabricator.services.mozilla.com/D94712)
* [Hardcode localhost to loopback](https://phabricator.services.mozilla.com/D92716)

> </details>
> 
> <details><summary>WebKit</summary>

Reviewer: ```suggestion
###  WebKit
```

* [Don't treat loopback addresses (127.0.0.0/8, ::1/128, localhost, .localhost) as mixed content](https://bugs.webkit.org/show_bug.cgi?id=171934)
* [Don't treat loopback IP addresses (127.0.0.0/8, ::1/128) as mixed content](https://bugs.webkit.org/show_bug.cgi?id=218623)
* [Introduce preference not to treat localhost and .localhost as mixed content](https://bugs.webkit.org/show_bug.cgi?id=218627)
* [[GTK] Allow WebKitTestServer to run non-loopback addresses for API test](https://bugs.webkit.org/show_bug.cgi?id=219257)
* [Migrate WebKitTestServer to libsoup 2.48 API](https://bugs.webkit.org/show_bug.cgi?id=219160)
* [Treat loopback addresses (127.0.0.0/8, ::1/128, localhost, .localhost) as potentially trustworthy URL ](https://bugs.webkit.org/show_bug.cgi?id=218980)

> </details>

Reviewer: ```suggestion
```

## In Conclusion

> IPFS and Igalia have made an initial effort to improve support and interoperability of web platform features that would benefit the distributed web, as well as the web community in general. In addition to starting discussions among the different actors, several patches have already landed in browsers. We are looking forward to continuing this work in 2021... Stay tuned!

Reviewer: obligatory rocket emoji 

```suggestion
IPFS and Igalia have made an initial effort to improve support and interoperability of web platform features that would benefit the distributed web, as well as the web community in general. In addition to starting discussions among the different actors, several patches have already landed in browsers. We are looking forward to continuing this work in 2021... Stay tuned! 🚀
```

<div style="width: 304pt; margin-left: auto; margin-right: auto;">
<!-- Generated by graphviz version 2.43.0 (0) -->
<!-- Pages: 1 -->
<svg width="304pt" height="305pt"
 viewBox="0.00 0.00 304.04 305.00" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 301)">
<polygon fill="transparent" stroke="transparent" points="-4,4 -4,-301 300.04,-301 300.04,4 -4,4"/>
<!-- 0 -->
<g id="node1" class="node">
<title>0</title>
<polygon fill="none" stroke="black" points="103.04,-297 49.04,-297 49.04,-261 103.04,-261 103.04,-297"/>
<text text-anchor="middle" x="76.04" y="-275.3" font-family="Times,serif" font-size="14.00">User</text>
</g>
<!-- 1 -->
<g id="node2" class="node">
<title>1</title>
<polygon fill="none" stroke="black" points="114.54,-210 37.54,-210 37.54,-174 114.54,-174 114.54,-210"/>
<text text-anchor="middle" x="76.04" y="-188.3" font-family="Times,serif" font-size="14.00">Browser</text>
</g>
<!-- 0&#45;&gt;1 -->
<g id="edge1" class="edge">
<title>0&#45;&gt;1</title>
<path fill="none" stroke="black" d="M76.04,-260.8C76.04,-246.05 76.04,-224.92 76.04,-210.18"/>

> <text text-anchor="middle" x="128.54" y="-231.8" font-family="Times,serif" font-size="14.00">ifps://bafkqaaa</text>

Reviewer: ```suggestion
<text text-anchor="middle" x="128.54" y="-231.8" font-family="Times,serif" font-size="14.00">ipfs://bafkqaaa</text>
```

* [2595424: Use a standard scheme to test potential trustworthiness](https://chromium-review.googlesource.com/c/chromium/src/+/2595424)
* [2593629: Add tests for SecurityOrigin::IsSecure and network::Is*PotentiallyTrustworthy](https://chromium-review.googlesource.com/c/chromium/src/+/2593629)
* [2610100: Add tests for SecurityOrigin::IsPotentiallyTrustworthy](https://chromium-review.googlesource.com/c/chromium/src/+/2610100)
* [2612947: Remove SecurityPolicy APIs for handling a trustworthy safelist](https://chromium-review.googlesource.com/c/chromium/src/+/2612947)
* [2614784: Remove SchemeRegistry APIs for handling local and secure schemes](https://chromium-review.googlesource.com/c/chromium/src/+/2614784)
* [2615260: Implement SecurityOrigin::IsPotentiallyTrustworthy with network::IsOriginPotentiallyTrustworthy](https://chromium-review.googlesource.com/c/chromium/src/+/2615260)
* [2617709: Rewrite SecurityOrigin::IsSecure() using GURL and url::Origin](https://chromium-review.googlesource.com/c/chromium/src/+/2617709)
* [2617883: Make SecurityOrigin::IsSecure treat localhost and local files as secure](https://chromium-review.googlesource.com/c/chromium/src/+/2617883)

> </details>
> 
> <details><summary>Firefox</summary>

Reviewer: ```suggestion

### Firefox
```
